DOC: Clarify isreal docstring

amrit110 opened the pull request:
Add a clarification to docstring of `isreal` method. Raised in #18814, `isreal` returns `True` for non-numeric inputs like `None`.

Closes #18814

amrit110 committed:
Add clarification to docstring for isreal

amrit110 committed:
DOC: add clarification to docstring for isreal

Add clarification to docstring of isreal function,
that it returns True for non-numeric inputs.
Closes #18814.

amrit110 committed:
Merge branch 'improve_docstring_isreal' of github.com:amrit110/numpy into improve_docstring_isreal

amrit110 committed:
Merge branch 'improve_docstring_isreal' of github.com:amrit110/numpy into improve_docstring_isreal

seberg commented on numpy/lib/type_check.py:
```diff
@@ -252,6 +252,10 @@ def isreal(x):
 
+    Returns True for non-numeric input elements, since the input is
+    interpreted to be an object array for which imaginary units of a complex 
+    number is not defined.
```
I am wondering if we should move this into the (currently not existing) `Notes` section.  Or alternatively mark with a `.. note::` markup if we want to make it a proper warning?  Or maybe both: Write in the notes section that a non-complex array returns all `True`, and then add a `.. note::` or `.. warning::` markup because for "object" that is a bit weird (if you have an object array containing complex numbers it is arguably wrong!)

The example is probably good, but should have a short comment before it to explain the context I think.

amrit110 replied:
Thanks for the pointers! I've modified the PR accordingly https://github.com/numpy/numpy/pull/18843/files.

amrit110 committed:
DOC: simplify clarification docstring for isreal

Add clarification to Notes section
Add clarification also as markup to serve as warning in docstring

amrit110 committed:
DOC: simplify clarification docstring for isreal

Add clarification to Notes section
Add clarification also as markup to serve as warning in docstring

amrit110 committed:
Merge branch 'improve_docstring_isreal' of github.com:amrit110/numpy into improve_docstring_isreal

charris commented:
You have a long line that needs to be < 80 characters.

amrit110 committed:
Limit line to 80 chars

amrit110 commented:
> You have a long line that needs to be < 80 characters.

@charris thanks for pointing out, fixed!

rgommers commented on numpy/lib/type_check.py:
```diff
@@ -252,6 +252,9 @@ def isreal(x):
     for that element is True.
 
+    .. note:: Returns all True for input elements which are not of complex
+    type.
```
This syntax won't work, can you change it to look like the `.. note::` directive in https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#paragraph-level-markup?

amrit110 replied:
ah!! thanks for pointing, fixed!

rgommers commented on numpy/lib/type_check.py:
```diff
@@ -262,6 +265,10 @@ def isreal(x):
         Boolean array of same shape as `x`.
 
+    Notes
+    --------
```
Underline needs to be the same length as the title.

amrit110 replied:
fixed!

rgommers commented on numpy/lib/type_check.py:
```diff
@@ -271,6 +278,10 @@ def isreal(x):
     >>> np.isreal([1+1j, 1+0j, 4.5, 3, 2, 2j])
     array([False,  True,  True,  True,  True, False])
+    
+    Returns True for all elements in input array which are not of complex type. 
```
You need a blank line between text and code examples here

amrit110 replied:
fixed!

amrit110 committed:
DOC: add fixes to docstring

cuspymd commented on numpy/lib/type_check.py:
```diff
@@ -271,6 +279,11 @@ def isreal(x):
+    
+    Returns True for all elements in input array which are not of complex type. 
+
+    >>> np.isreal([None, False, "imaginary", complex])
+    array([True, True, True, True])
```
When tested, it returns False instead of an array in the following cases.
```
>>> np.isreal([2j, "a"])
/.../lib/python3.6/site-packages/numpy/lib/type_check.py:277: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  return imag(x) == 0
False

```

seberg replied:
That happens for string arrays, on current master, it should run into two future warnings, because the original input is already a mix of string and complex... Maybe fine to ignore.

seberg replied:
Maybe we should keep the `Notes` section as it is, and move the `.. note::` directive down there with more details that the function does not work for string arrays, and for `dtype=object` currently always returns True even if the content is complex.

If we want an example, maybe include `np.array([1+2j, ...], dtype=object)` (a complex number and the `dtype=object` explicitly)?

amrit110 replied:
@cuspymd, @seberg, thanks for the comments. i've made fixes by moving the `.. note::` directive down to the examples section as well as including the example for string array and mentioning that the function doesn't work, as well as the array with python objects returning all `True` even if it consists complex numbers. let me know if more clarifications can be added!

cuspymd commented on numpy/lib/type_check.py:
```diff
@@ -271,6 +279,11 @@ def isreal(x):
     >>> np.isreal([1+1j, 1+0j, 4.5, 3, 2, 2j])
     array([False,  True,  True,  True,  True, False])
+    
+    Returns True for all elements in input array which are not of complex type. 
```
"If there is at least one element of the input array that is not of numeric types" Isn't this the explanation you intended?

amrit110 replied:
yep, now its clearer since the example creates an array of dtype=object, and ive clarified that it returns True for all elements if so

amrit110 committed:
DOC: improve docstring, clarifying examples

rossbar commented on numpy/lib/type_check.py:
```diff
@@ -262,6 +262,10 @@ def isreal(x):
     out : ndarray, bool
         Boolean array of same shape as `x`.
 
+    Notes
+    -----
+    Returns all True for non-complex array.
+
```
IMO this statement isn't specific and doesn't clarify anything, so I'd just as soon remove it. Alternatively, the strange behavior for certain dtypes could be highlighted:

```suggestion
    Notes
    -----
    `isreal` may behave unexpectedly for string or object arrays (see examples)
```

amrit110 commented on numpy/lib/type_check.py:
```diff
@@ -271,6 +275,19 @@ def isreal(x):
+    >>> np.isreal([2j, "a"])
+    False
+    
+    Returns True for all elements in input array of `dtype=object` even if 
+    any of the elements is complex.
+
+    >>> np.array([1+2j, ...], dtype=object)
+    array([True,  True])
```
@rossbar ahh missed showing the isreal call in the next line, thanks for the suggested fix! Also agree that making it explicitly clear that for object array even if all elements are complex it still returns True

amrit110 committed:
Update numpy/lib/type_check.py

fix to docstring

Co-authored-by: Ross Barnowski <rossbar@berkeley.edu>

amrit110 committed:
Update numpy/lib/type_check.py

improve note to explicitly warn about string or object arrays

Co-authored-by: Ross Barnowski <rossbar@berkeley.edu>

amrit110 commented on numpy/lib/type_check.py:
```diff
@@ -271,6 +275,19 @@ def isreal(x):
     --------
     >>> np.isreal([1+1j, 1+0j, 4.5, 3, 2, 2j])
     array([False,  True,  True,  True,  True, False])
+    
+    .. note::
```
agreed!

amrit110 committed:
Update numpy/lib/type_check.py

remove note directive in examples section

Co-authored-by: Ross Barnowski <rossbar@berkeley.edu>

amrit110 committed:
DOC: improve docstring clarifications

amrit110 commented on numpy/lib/type_check.py:
```diff
@@ -271,6 +275,19 @@ def isreal(x):
+    The function does not work on string arrays.
+
+    >>> np.isreal([2j, "a"])
+    False
```
hmm, yeah i think it seems worthwhile to show the mixed case as an example now that the notes warn of unexpected behaviour for string and object arrays. thanks for pointing out the explicit conversion to unicode string.

amrit110 commented on numpy/lib/type_check.py:
```diff
@@ -271,6 +275,25 @@ def isreal(x):
     --------
     >>> np.isreal([1+1j, 1+0j, 4.5, 3, 2, 2j])
```
@rossbar i wonder if i should leave this as is or create an array and apply the function on the array for consistency. if latter, something like `a = np.isreal([1+1j, 1+0j, 4.5, 3, 2, 2j], dtype='c16')` and then apply `np.isreal(a)`

rossbar replied:
I agree that being explicit about dtypes when creating arrays is generally a good practice; though this case is straightforward. I'd say it's up to you :) - though I think you meant `np.array` instead of `np.isreal` in your suggestion, and I would use the more generic `dtype=complex`.

amrit110 replied:
yep, thanks!

amrit110 committed:
DOC: update docstring example for consistency

rossbar committed:
rst formatting - inline literal.

rossbar reviewed (approved):
LGTM, thanks for sticking with all the iterations @amrit110 !

rossbar commented:
CI failures are unrelated - the doctests in the circleci job are green and the [rendered page](https://19658-908607-gh.circle-artifacts.com/0/doc/build/html/reference/generated/numpy.isreal.html) looks good.

amrit110 commented:
thanks @rossbar ! can i merge it or is to be done by one of the maintainers?
